### ISSUE TITLE
Allow for switching between Orders and Subscriptions

### ISSUE THREAD

**Issue by JasonTheAdams** [paypal/paypal-sdk-client#44]
Greetings!

I work with [givewp.com](https://givewp.com) and we've been working with an integration manager at PayPal since June to provide integration for PayPal Commerce.

We've recently run into a problem with this SDK wherein the `vault=true` parameter must be set at the time that the SDK script is loaded. This means that the SDK is locked into either creating Orders or Subscriptions once it's loaded.

This is a problem for us as our product provides organizations with a donation form to put on their website. During realtime, the donor may decide whether they're making a one-time or recurring donation. In fact, the donor could even start willing out CC details, then decide they want to make it recurring, tick a box, and continue. You can see a couple demos:
- https://givewp.com/demos/recurring-donations/
- https://demo.givewp.com/donate-2/

Is there currently a way of switching between Orders and Subscriptions at runtime? If not, we need that for compatibility of integration.

Thank you!

**Comment by gregjopa** [comment]
> Is there currently a way of switching between Orders and Subscriptions at runtime?

Hi @JasonTheAdams, the JS SDK can be reloaded on the page for handling query parameter changes. Here's a codepen to demonstrate: https://codepen.io/gregjopa/pen/jOqBGJy. 

<img width="723" alt="Screen Shot 2020-08-25 at 2 06 38 PM" src="https://user-images.githubusercontent.com/534034/91216850-371c0c00-e6dc-11ea-8e58-403daf5c9482.png">

This demo is using [paypal-js](https://github.com/paypal/paypal-js) to simplify the loading of the sdk script. At the end of the day this is just inserting the new sdk script like so `document.head.insertBefore(newScript, document.head.firstElementChild)` which will load and take priority and cleanup the old sdk script on the page.

Would this proposed solution work for your use case?

**Comment by JasonTheAdams** [comment]
Thanks, @gregjopa! I appreciate you taking the time to put that together! Obviously, not having to reload the script and re-render — causing the flash — would be ideal for a future feature. But this seems like it could be a solid workaround.

We'll give this a shot this week and see if it works within our system and I'll report back to this issue with our success.

Thank you!

**Comment by gregjopa** [comment]
@JasonTheAdams happy to help! I'll leave this issue open. Please let us know how the integration goes.

**Comment by ravinderk** [comment]
@gregjopa When I define a `components` property of PayPal SDK then click on the radio button does not overwrite old button and the number of button count keeps increase with a javascript error. Can you review this pen: https://codepen.io/ravinderk/pen/wvGevRg


![image](https://user-images.githubusercontent.com/1784821/91576992-bbcd8b00-e965-11ea-833b-9fa4409d2f5b.png)

![paypa-sdk](https://user-images.githubusercontent.com/1784821/91577564-84131300-e966-11ea-9210-b2009dfd59a6.gif)

![image](https://user-images.githubusercontent.com/1784821/91577253-136bf680-e966-11ea-9aee-0814ff39bc3e.png)

**Comment by gregjopa** [comment]
@ravinderk that is by design. Our motto is you can never have too many paypal buttons on a web page.

Just kidding 😆. To avoid showing multiple paypal buttons, you need to clean up the old before rendering the new. There's a helpful `close()` function to doing this cleanup work. Ex:

```js
  let buttons;

  function render(options) {
    if (buttons) {
      buttons.close();
    }
    buttons = paypal.Buttons(options);
    buttons.render(document.querySelector('#paypal-button-container'));
  }
```

Here's a pen demoing this cleanup behavior: https://codepen.io/gregjopa/pen/RwagNWR?editors=1000

**Comment by ravinderk** [comment]
@gregjopa I am still seeing this error. Can you check it?

![image](https://user-images.githubusercontent.com/1784821/91604517-69ec2b80-e98c-11ea-8b15-f2814350c036.png)

**Comment by gregjopa** [comment]
@ravinderk That first error seems to related to hosted fields. The error does not happen when using `components=buttons` for both cases. I updated the codepen to reflect this: https://codepen.io/gregjopa/pen/RwagNWR?editors=1000.

My guess is you also need to cleanup hosted fields in the same manner as the buttons:
```js
if (fields) {
  fields.close();
}

fields = window.paypal.hostedFields({ 
  // your config here 
});
```

And the second error seems to be a side effect of the first error.

**Comment by ravinderk** [comment]
@gregjopa Is `window.paypal.hostedFields` is a function? How will I get access to it?

![image](https://user-images.githubusercontent.com/1784821/91606758-4aef9880-e990-11ea-869b-83db391cec4c.png)

**Comment by gregjopa** [comment]
@ravinderk whoops I was wrong. HostedFields is not a function. It's an object and your screenshot is correct. You will need to use it's `render()` function like so: 

```js
paypal.HostedFields.render({ /* ... */ })
    .then(function (hostedFieldsInstance) { /* ... */ })
```

https://developer.paypal.com/docs/business/checkout/reference/style-guide/#style-reference-for-advanced-credit-and-debit-card-payments. 

Note that hostedFields is kind of off topic with this issue. Please open up a new issue if you have additional questions about implementing hostedFields.

**Comment by JasonTheAdams** [comment]
@gregjopa From our perspective this is all the same thing. We're implementing _both_ the Smart Buttons and Advanced Fields within our forms. We're not having a problem with this combination except in the context of switching between one-time and recurring donations.

We could rename this Issue to "Allow for switching Smart Buttons between Orders and Subscriptions" and then create another "Allow for switching Hosted Fields between Orders and Subscriptions", if you would like. For us, though, we can't proceed until both are working.

**Comment by gregjopa** [comment]
@JasonTheAdams thanks for clarifying. I didn't realize you and @ravinderk are working together 😄 Let's keep this one issue. 
I'm checking with the team about how to clean up hosted-fields. I'll follow up soon.

**Comment by JasonTheAdams** [comment]
Oh yeah, I guess there's no reason you'd have known that. Hahah! Thank you! 😄

**Comment by JasonTheAdams** [comment]
Hey @gregjopa! Any word on this? Just checking in. Thanks!

**Comment by gregjopa** [comment]
Hey @JasonTheAdams, thanks for your patience. We've reviewed this issue internally and there is a bug with hosted-fields "destroy" hook for the reload use case. A fix has been submitted and should be released by the end of next week. 

I've updated this [codepen example](https://codepen.io/gregjopa/pen/RwagNWR?editors=0010) for verifying this fix. Currently it throws the following error when switching between Orders and Subscriptions: 
> Uncaught Error: Request listener already exists for zoid_allow_delegate_payments_sdk_contingency_handler on domain * for wildcard window

This error should go away after the fix is released next week. I'll follow up once the fix is in prod.

**Comment by JasonTheAdams** [comment]
Roger that! Thanks for the update and pending fix!

**Comment by gregjopa** [comment]
> Uncaught Error: Request listener already exists for zoid_allow_delegate_payments_sdk_contingency_handler on domain * for wildcard window

Hi @JasonTheAdams and @ravinderk, I'm happy to share that the above error has been fixed in yesterday's v5.0.164 release!

Thanks for bringing this to our attention. I'm going to close this one out.

**Comment by JasonTheAdams** [comment]
Thank you, @gregjopa! We appreciate you and your team taking the time to look into this and resolve the issue. It will be a few weeks, likely, before we're able to implement this, but from the looks of your codepen it seems to be working great. 👍

**Comment by ravinderk** [comment]
@gregjopa I find out that subscription with smart button works fine even I set `intent` to `capture`. `subscription` value is not mentioned in allowed values: https://developer.paypal.com/docs/checkout/reference/customize-sdk/#intent

I able to process subscription but I see a warning in Chrome `console` tab.
![image](https://user-images.githubusercontent.com/1784821/94439308-dfcaf900-01bd-11eb-9046-2b681b4f6376.png)

- Are you planning to make `intent=subscirption` required in the future? 
- I want to mention at the same time webpage can contain two forms that can process subscription and order both., so I am trying to find a way to skip reloading the script just for the `intent` param.
- `vault` param is required for a subscription but I can also process the one-time payment with it. Am I missing something here?

As per testing, things working fine but I am trying to make sure that the solution is future proof.
